doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,249 @@
+\ingroup module_hidden
+
+\page goto-program-transformations Goto Program Transformations
+
+TODO:
+- Add link for each pass to the function that implements it.
+- Add to each title the name of the function that implements it.
+- Add link points and cross-linking
+- Fix consistency of Removal/Remove etc.
+
+## Core Transformation Passes
+
+This section lists the transformation passes that must have been applied for the
+goto model to be in core goto format.
+
+Note that the passes are listed below in the order they are currently called in
+CBMC. While all dependencies on the ordering are not fully documented, the
+following order should be used.
+
+### Removal/Lowering of Assembly
+
+This pass goes through the goto model and removes or lowers instances of
+assembly intructions. Assembly instructions are stored in instances of the
+`other` instruction.
+
+_This pass has no predeccesor._
+
+### String Instrumentation
+
+This (optional) pass adds checks on calls to some of the C standard library
+string functions. It uses the tracking information added by the
+[string-abstraction] pass. It is switched on by the `--string-abstraction`
+command line option.
+
+_Predecessor passes are [Removal/Lowering of Assembly] or
+[Linking of Standard Libraries]._
+
+### Removal/Lowering of Function Pointers
+
+This pass finds all the instructions which are function calls to the value of a
+function pointer. Each instruction is then replaced with a switch block. The
+switch block contains a case for each of the potential targets of the  function
+pointer. The targets are found by looking for all functions in the goto program
+that match the type of the function pointer.
+
+Note that this pass must be applied after all linking has been done. Either for
```

Comment:
The requirement is checked by an assertion that the fall through is unreachable. See - https://github.com/diffblue/cbmc/blob/d7099aeaff5043b65aba669fe221cc9b9af40c8c/src/goto-programs/remove_function_pointers.cpp#L435

I may have worded this a little too strongly. If you can show via data flow analysis that all potential targets of a function pointer at the point it is called are included in the model then applying this pass to that function call would yield a good result. Indeed this could result in a faster analysis due to the smaller set of functions being included in the switch block.

The current approach this pass uses has the drawback that the more functions which are linked into the model the larger the function table generated may be, but the earlier it is done in terms of linking then the more chance that valid targets are not included. Resolving this would require rewriting it using an entirely different approach. Two  potential solutions I can think of would be -

1. Make the lowering part of symex. This would be a fairly complete reworking of the current implementation and would add to the complexities of symex.
2. Use the Abstract Interpreter with variable sensitivity to build the value sets of function pointers. This would require fixing the known issues which the current implementation has with recursive function calls.